jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java

```diff
@@ -812,173 +628,55 @@ protected void parseDir(Set<? extends Handler> handlers, Resource root) throws E
             return;
 
-        if (FileID.isJavaArchive(jarResource.getPath()))
-        {
-            if (LOG.isDebugEnabled())
-                LOG.debug("Scanning jar {}", jarResource);
-
-            ExceptionUtil.MultiException multiException = new ExceptionUtil.MultiException();
-            try (MultiReleaseJarFile jarFile = new MultiReleaseJarFile(jarResource.getPath());
-                 Stream<Path> jarEntryStream = jarFile.stream()
-                     .filter(Files::isRegularFile)
-                     .filter(FileID::isNotModuleInfoClass)
-                     .filter(FileID::isNotMetaInfVersions)
-                     .filter(FileID::isClassFile)
-            )
-            {
-                jarEntryStream.forEach(e ->
-                {
-                    try
-                    {
-                        parseJarEntry(handlers, jarResource, e);
-                    }
-                    catch (Exception ex)
-                    {
-                        multiException.add(new RuntimeException("Error scanning entry " + e.toString(), ex));
-                    }
-                });
-            }
-            multiException.ifExceptionThrow();
-        }
-    }
-
-    /**
-     * Parse a single entry in a jar file
-     *
-     * @param handlers the handlers to look for classes in
-     * @param entry the entry in the potentially MultiRelease jar resource to parse
-     * @param jar the jar file
-     * @throws Exception if unable to parse
-     */
-    protected void parseJarEntry(Set<? extends Handler> handlers, Resource jar, Path entry)
-        throws Exception
-    {
-        if (jar == null || entry == null)
-            return;
-
-        //skip directories
-        if (Files.isDirectory(entry))
+        if (!FileID.isJavaArchive(jarResource.getPath()))
             return;
 
-        String name = entry.toString();
+        if (LOG.isDebugEnabled())
+            LOG.debug("Scanning jar {}", jarResource);
 
-        //check file is a valid class file name
-        if (isValidClassFileName(name) && isValidClassFilePath(name))
+        Path jarPath = jarResource.getPath();
+        try (Resource.Mount jarMount = Resource.mountJar(jarPath))
         {
-            String shortName = StringUtil.replace(name, '/', '.').substring(0, name.length() - 6);
-            URI location = entry.toUri();
-            addParsedClass(shortName, location);
-
-            try (InputStream is = Files.newInputStream(entry))
-            {
-                if (LOG.isDebugEnabled())
-                    LOG.debug("Scanning class from jar {}", location);
-                LOG.debug("Scanning class from jar {}", location);
-                scanClass(handlers, jar, is);
-            }
+            Path root = jarMount.root().getPath();
+            parseDir(handlers, jarResource, root);
         }
     }
 
+    private void parseClassFile(Set<? extends Handler> handlers, Resource jar, Path root, Path entry) throws IOException
```

Comment:
I don't think there should be both `parseClassFile` and `parseClass` methods, should just be one method.

Reply:
There's a small difference.

The one with `Path root, Path entry` allows the "shortname" to be calculated properly.
Which takes something like ...

```
 root = /path/to/scan/
entry = /path/to/scan/org/eclipse/jetty/demo/Foo.class
```

And is able to create a proper shortname of `org.eclipse.jetty.demo.Foo`, which is used in the `_parsedClassNames` Map.
However, the raw/naked scan of a Path, without a root (something we used to support) difficult to generate a shortName from.

See Jetty 10 -> https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-annotations/src/main/java/org/eclipse/jetty/annotations/AnnotationParser.java#L741-L748

This code path cannot ever generate a proper classname.
So the old Jetty 10 behavior was to not track that class.
This old code path (raw class file parse) is the only reason for this difference in methods.
If you are ok with eliminating this code path, then we can simplify this difference in parseClass* methods.

Reply:
Turns out using the ASM parse results for the classname is a better choice, much easier code.